### PR TITLE
[SVS-9] Refactor existing code

### PR DESCRIPTION
**Pull request by eyalantzMS**
Refactor code to support detection of unbound project that will use some of the existing code that needed to shared with the new functionality.

Specifically:
1. Need to load rule sets in a safe way (we had code that does it so I pushed that into the serializer).
2. Need to get the solution rule sets folder (extracted that into a service method that supports open/closed solutions).
3. Need to find the rule set project absolute file path (extracted that into a service method that will verify that the file is on disk).
4. Need to find all the solution rule sets includes - we already had a method that had that part of it, just extracted that functionality into a separate method.

Misc: moved various MEF services and MEF service interfaces into MefServices folder

### CODE REVIEW

File: src/Integration.UnitTests/Binding/ProjectBindingOperation.WriterTests.cs [hunk 1]
```diff
@@ -20,45 +20,6 @@ public partial class ProjectBindingOperationTests
         #region Tests
 
         [TestMethod]
-        public void ProjectBindingOperation_SafeLoadRuleSet()
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/Binding/ProjectBindingOperation.WriterTests.cs]: Responsibility (and test) moved into the RuleSetSerialzier

File: src/Integration.UnitTests/LocalServices/SolutionRuleSetsInformationProviderTests.cs [hunk 2]
```diff
@@ -166,15 +164,120 @@ public void SolutionRuleSetsInformationProvider_CalculateSolutionSonarQubeRuleSe
...
+            // Case 2: C# + valid path characters
+            // Act
+            ruleSetPath = testSubject.CalculateSolutionSonarQubeRuleSetFilePath("MyKey", RuleSetGroup.CSharp);
+
+            // Verify
+            Assert.AreEqual(@"z:\folder\solution\SonarQube\MyKeyCSharp.ruleset", ruleSetPath);
+        }
+
+        [TestMethod]
+        public void SolutionRuleSetsInformationProvider_CalculateSolutionSonarQubeRuleSetFilePath_OnClosedSolution()
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/LocalServices/SolutionRuleSetsInformationProviderTests.cs]: Previously missing test case.

File: src/Integration/LocalServices/RuleSetSerializer.cs [hunk 3]
```diff
@@ -6,19 +6,67 @@
...
+                    // Log this for testing purposes
+                    Trace.WriteLine(ex.ToString(), nameof(LoadRuleSet));
+                }
+            }
+
+            return null;
+
+        }
+
+        public void WriteRuleSetFile(RuleSet ruleSet, string ruleSetPath)
```
Comment (eyalantzMS) [inline src/Integration/LocalServices/RuleSetSerializer.cs]: Writing is not safe, we're handing and output the exception when it's using (part of binding)

File: src/Integration/LocalServices/SolutionRuleSetsInformationProvider.cs [hunk 4]
```diff
@@ -58,26 +58,57 @@ public IEnumerable<RuleSetDeclaration> GetProjectRuleSetsDeclarations(Project pr
...
             if (string.IsNullOrWhiteSpace(sonarQubeProjectKey))
             {
                 throw new ArgumentNullException(nameof(sonarQubeProjectKey));
             }
 
-            if (string.IsNullOrWhiteSpace(fileNameSuffix))
+            string ruleSetDirectoryRoot = this.GetSolutionSonarQubeRulesFolder();
+
+            if (string.IsNullOrWhiteSpace(ruleSetDirectoryRoot))
             {
```
Comment (eyalantzMS) [inline src/Integration/LocalServices/SolutionRuleSetsInformationProvider.cs]: Added a test case for this.

File: src/Integration/LocalServices/SolutionRuleSetsInformationProvider.cs [hunk 5]
```diff
@@ -58,26 +58,57 @@ public IEnumerable<RuleSetDeclaration> GetProjectRuleSetsDeclarations(Project pr
...
-            var projectSystem = this.serviceProvider.GetService<IProjectSystemHelper>();
-            string solutionFullPath = projectSystem.GetCurrentActiveSolution().FullName;
-
-            string solutionRoot = Path.GetDirectoryName(solutionFullPath);
-            string ruleSetDirectoryRoot = Path.Combine(solutionRoot, Constants.SonarQubeManagedFolderName);
+            string fileNameSuffix = ruleSetGroup.ToString();
             return GenerateSolutionRuleSetPath(ruleSetDirectoryRoot, sonarQubeProjectKey, fileNameSuffix);
         }
 
+        public bool TryGetProjectRuleSetFilePath(Project project, RuleSetDeclaration declaration, out string fullFilePath)
```
Comment (eyalantzMS) [inline src/Integration/LocalServices/SolutionRuleSetsInformationProvider.cs]: The code existed before, I just moved it here

File: src/Integration/ProfileConflicts/ConflictsManager.cs [hunk 6]
```diff
@@ -110,8 +110,9 @@ private RuleSetInformation[] GetAggregatedSolutionRuleSets()
 
             foreach (Project project in projectSystem.GetFilteredSolutionProjects())
             {
-                string suffix = SolutionBindingOperation.GetProjectRuleSetSuffix(ProjectBindingOperation.GetProjectGroup(project));
```
Comment (eyalantzMS) [inline src/Integration/ProfileConflicts/ConflictsManager.cs]: This bit is now common and an implementation detail of the service.

File: src/Integration/RuleSetGroup.cs [hunk 7]
```diff
@@ -5,7 +5,7 @@
 // </copyright>
 //-----------------------------------------------------------------------
 
-namespace SonarLint.VisualStudio.Integration.Binding
+namespace SonarLint.VisualStudio.Integration
```
Comment (eyalantzMS) [inline src/Integration/RuleSetGroup.cs]: Not specific to just binding anymore.

File: src/Integration/RuleSetHelper.cs [hunk 8]
```diff
@@ -43,23 +43,34 @@ public static void UpdateExistingProjectRuleSet(RuleSet ruleSet, string solution
         }
 
         /// <summary>
-        /// Remove all rule set inclusions which exist under the specified root directory.
+        /// Find all the <see cref="RuleSetInclude"/>, for a <paramref name="ruleSet"/>, 
+        /// which are referencing rule sets under <paramref name="rootDirectory"/>
         /// </summary>
-        internal /* testing purposes */ static void RemoveAllIncludesUnderRoot(RuleSet ruleSet, string rootDirectory)
+        public static IEnumerable<RuleSetInclude> FindAllIncludesUnderRoot(RuleSet ruleSet, string rootDirectory)
```
Comment (eyalantzMS) [inline src/Integration/RuleSetHelper.cs]: Existed before, extracted to a separate method.

File: src/Integration/LocalServices/SolutionRuleSetsInformationProvider.cs [hunk 9]
```diff
@@ -58,26 +58,57 @@ public IEnumerable<RuleSetDeclaration> GetProjectRuleSetsDeclarations(Project pr
             }
         }
 
-        public string CalculateSolutionSonarQubeRuleSetFilePath(string sonarQubeProjectKey, string fileNameSuffix)
+        public string GetSolutionSonarQubeRulesFolder()
+        {
+            var projectSystem = this.serviceProvider.GetService<IProjectSystemHelper>();
+            string solutionFullPath = projectSystem.GetCurrentActiveSolution().FullName;
```
Comment (mjcheetham) [inline src/Integration/LocalServices/SolutionRuleSetsInformationProvider.cs]: `GetCurrentActiveSolution()` could return `null` (in the case there is no active solution). Use `GetCurrentActiveSolution()?.FullName`

File: src/TestInfrastructure/Framework/ConfigurableSolutionRuleSetsInformationProvider.cs [hunk 10]
```diff
@@ -31,11 +30,24 @@ IEnumerable<RuleSetDeclaration> ISolutionRuleSetsInformationProvider.GetProjectR
...
+        string ISolutionRuleSetsInformationProvider.GetSolutionSonarQubeRulesFolder()
+        {
+            return Path.Combine(this.SolutionRootFolder, Constants.SonarQubeManagedFolderName);
+        }
 
-        string ISolutionRuleSetsInformationProvider.CalculateSolutionSonarQubeRuleSetFilePath(string sonarQubeProjectKey, string fileNameSuffix)
+        string ISolutionRuleSetsInformationProvider.CalculateSolutionSonarQubeRuleSetFilePath(string sonarQubeProjectKey, RuleSetGroup group)
         {
-            return Path.Combine(this.SolutionRootFolder, Constants.SonarQubeManagedFolderName, sonarQubeProjectKey + fileNameSuffix + "." + Constants.RuleSetFileExtension);
+            string fileName = sonarQubeProjectKey + group.ToString() + "." + Constants.RuleSetFileExtension;
```
Comment (mjcheetham) [inline src/TestInfrastructure/Framework/ConfigurableSolutionRuleSetsInformationProvider.cs]: Would `$"{sonarQubeProjectKey}{group}.{Constants.RuleSetFileExtension}";` not be clearer to see the structure?

File: src/Integration.UnitTests/VsSessionHostTests.cs [hunk 11]
```diff
@@ -11,6 +11,7 @@
 using SonarLint.VisualStudio.Integration.TeamExplorer;
 using SonarLint.VisualStudio.Integration.WPF;
 using System;
+using System.ComponentModel.Design;
```
Comment (mjcheetham) [inline src/Integration.UnitTests/VsSessionHostTests.cs]: Why?

**Comment by mjcheetham** [comment]
Comments aside, LGTM

File: src/Integration.UnitTests/VsSessionHostTests.cs (hunk 11 above)
Comment (eyalantzMS) [inline src/Integration.UnitTests/VsSessionHostTests.cs]: Not needed anymore

File: src/TestInfrastructure/Framework/ConfigurableSolutionRuleSetsInformationProvider.cs (hunk 10 above)
Comment (eyalantzMS) [inline src/TestInfrastructure/Framework/ConfigurableSolutionRuleSetsInformationProvider.cs]: yep

File: src/Integration/LocalServices/SolutionRuleSetsInformationProvider.cs (hunk 9 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/SolutionRuleSetsInformationProvider.cs]: sure

**Pull request merged**